### ISSUE TITLE
Supports compiling casts between pointers of different types

### ISSUE THREAD

**Issue by psolaimani** [bedatadriven/renjin#276]
XVector examples fail due to GCC issue

here is the error, and here is the [full stack](http://packages.renjin.org/package/org.renjin.bioconductor/XVector/0.12.1/build/13#test-intra-range-methods-examples)
```
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.renjin.invoke.reflection.FunctionBinding$Overload.invoke(FunctionBinding.java:117)
	... 173 more
Caused by: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    org/renjin/bioconductor/XVector/SharedInteger_class__.SharedInteger_read_ints_from_i1i2(Lorg/renjin/sexp/SEXP;Lorg/renjin/sexp/SEXP;Lorg/renjin/sexp/SEXP;)Lorg/renjin/sexp/SEXP; @173: invokestatic
  Reason:
    Type 'org/renjin/gcc/runtime/IntPtr' (current frame, stack[4]) is not assignable to 'org/renjin/gcc/runtime/BytePtr'
  Current Frame:
    bci: @173
    flags: { }
    locals: { 'org/renjin/sexp/SEXP', 'org/renjin/sexp/SEXP', 'org/renjin/sexp/SEXP', integer, integer, integer, 'org/renjin/sexp/SEXP', 'org/renjin/sexp/SEXP', top, '[I', integer, integer, top, '[I', integer, integer, top, top, top, '[I', integer, top, '[I', integer, 'org/renjin/gcc/runtime/IntPtr', 'org/renjin/gcc/runtime/IntPtr', 'org/renjin/gcc/runtime/IntPtr', 'org/renjin/gcc/runtime/IntPtr' }
    stack: { integer, integer, 'org/renjin/gcc/runtime/IntPtr', integer, 'org/renjin/gcc/runtime/IntPtr', integer, integer }
  Bytecode:
    0000000: 2ab8 0011 3a07 2bb8 001b 3a18 1918 b400
    0000010: 213a 1619 18b4 0025 3617 1916 1517 2e02
    0000020: 6036 052c b800 1b3a 1919 19b4 0021 3a13
    0000030: 1919 b400 2536 1419 1315 142e 0260 3604
    0000040: 1504 1505 6404 603e 100d 1db8 0043 3a06
    0000050: 1906 b800 4657 1907 b800 1736 0f19 07b8
    0000060: 001b 3a1a 191a b400 213a 0d19 1ab4 0025
    0000070: 360e 1906 b800 1736 0b19 06b8 001b 3a1b
    0000080: 191b b400 213a 0919 1bb4 0025 360a 1505
    0000090: 1504 bb00 1d59 1909 150a b700 2815 0bbb
    00000a0: 001d 5919 0d15 0eb7 0028 150f 07b8 0056
    00000b0: 04b8 004d 1906 b0                      

	at org.renjin.bioconductor.XVector.XVector.SharedInteger_new(Unknown Source)
	... 178 more
```

**Comment by akbertram** [comment]
This is caused by the C code in [SharedInteger_class.c](https://github.com/Bioconductor-mirror/XVector/blob/release-3.3/src/SharedInteger_class.c#L63):

```.c
	_Ocopy_byteblocks_from_i1i2(i1, i2,
			(char *) INTEGER(tag), LENGTH(tag),
			(char *) INTEGER(src_tag), LENGTH(src_tag), sizeof(int));
```

Which casts an `int*` to a `byte*`. Currently GCC-Bridge does not support this.

**Comment by psolaimani** [comment]
Cast necessary for IRanges:
org.renjin.gcc.codegen.fatptr.FatPtrPair to org.renjin.gcc.codegen.type.record.unit.RecordUnitPtr
org.renjin.gcc.runtime.ObjectPtr to org.renjin.bioconductor.IRanges.record$int_ae

**Comment by akbertram** [comment]
This has been resolved by #332. XVector now compiles, but most tests are still failing for unrelated reasons.
